### PR TITLE
Add tests for bios dependencies

### PR DESCRIPTION
**Pull request by pepeiborra**
I ran into issues trying to define component dependencies in my Bios cradles, and wrote a pair of failing tests to demonstrate. 
Please check that I'm not missing anything, and let me know if there is an obvious fix.

### CODE REVIEW

**Comment by fendor** [comment]
I can reproduce the issue. The "problem" is this call: https://github.com/mpickering/hie-bios/blob/master/src/HIE/Bios/Cradle.hs#L354
It essentially assumes that the files are written to the file pointed to by "HIE_BIOS_OUTPUT". I assume I missed to document this properly with the documentation overhaul.
So, this fix your issue, this works for me:
```yaml
cradle:
  bios:
    shell: |
      :; echo "-Wall" >> $HIE_BIOS_OUTPUT
      :; echo "A" >> $HIE_BIOS_OUTPUT
      :; echo "B" >> $HIE_BIOS_OUTPUT
      :; exit 0
      ECHO "-Wall" >> %HIE_BIOS_OUTPUT%
      ECHO "A" >> %HIE_BIOS_OUTPUT%
      ECHO "B" >> %HIE_BIOS_OUTPUT%
    dependency-shell: |
      :; echo "hie.yaml" >> $HIE_BIOS_OUTPUT
      :; exit 0
      ECHO "hie.yaml" >> %HIE_BIOS_OUTPUT%
```
```bash
> hie-bios debug A.hs
Root directory:        /home/munin/Documents/haskell/hie-bios/tests/projects/simple-bios-shell
Component directory:   /home/munin/Documents/haskell/hie-bios/tests/projects/simple-bios-shell
GHC options:           -Wall A B
GHC library directory: CradleSuccess "/nix/store/vhpfihshqj6sq5f9m2825gzz97nb3qkf-ghc-8.8.3/lib/ghc-8.8.3"
GHC version:           CradleSuccess "8.8.3"
Config Location:       /home/munin/Documents/haskell/hie-bios/tests/projects/simple-bios-shell/hie.yaml
Cradle:                Cradle {cradleRootDir = "/home/munin/Documents/haskell/hie-bios/tests/projects/simple-bios-shell", cradleOptsProg = CradleAction: Bios}
Dependencies:          hie.yaml
```

Do you think it is enough to document this or do you prefer writing thinks to stdout?

**Comment by pepeiborra** [comment]
Fixing the README is enough for me. Thanks for looking into this!

**Comment by pepeiborra** [comment]
Actually, there is another issue for me. The documentation says that `dependency-program` is called with no arguments, but that is inconsistent with `program` and doesn't really work for us. Is that by accident or by design?

**Comment by pepeiborra** [comment]
how do I fix the Windows tests here?

**Comment by fendor** [comment]
Afaict, the windows tests are just broken. Unfortunately, I dont know why they are broken, but you dont have to fix the tests right now, since HEAD is broken, too. 
I will debug them later today and report back.

**Comment by fendor** [comment]
So, after some digging, it seems like the polyglot bios shell script does not work on CI. Did you test it on windows?
I think that, because in the PR #236, in the latest commit, the test has the following fail message:
```
   simple-bios-shell:                             FAIL (0.30s)
      Finding Cradle for: C:\Users\runneradmin\AppData\Local\Temp\hie-bios-test-6313e53742483925\B.hs
      Loading Cradle: Just "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\hie-bios-test-6313e53742483925\\hie.yaml"
      Get runtime GHC library directory                              (0.07s)
      Get runtime GHC version                                        (0.05s)
      Initialise Flags                                               (0.16s)
      Initial module load
      comp opts: ComponentOptions {componentOptions = ["\"-Wall\" "], componentRoot = "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\hie-bios-test-6313e53742483925", componentDependencies = ["\"hie.yaml\" "]} (0.01s)
        target ‘"-Wall" ’ is not a module name or a source file
```

When the hie.yaml file looks like this:
```yaml
cradle:
  bios:
    shell: |
      ECHO "-Wall" >> %HIE_BIOS_OUTPUT%
      ECHO "A" >> %HIE_BIOS_OUTPUT%
      ECHO "B" >> %HIE_BIOS_OUTPUT%
    dependency-shell: |
      ECHO "hie.yaml" >> %HIE_BIOS_OUTPUT%
```

otherwise, the test-output is like this:
```
   simple-bios-shell:                             OK (0.55s)
      Finding Cradle for: C:\Users\runneradmin\AppData\Local\Temp\hie-bios-test-bd77a1b14cf6e038\B.hs
      Loading Cradle: Just "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\hie-bios-test-bd77a1b14cf6e038\\hie.yaml"
      Get runtime GHC library directory                              (0.06s)
      Get runtime GHC version                                        (0.03s)
      Initialise Flags                                               (0.16s)
      Initial module load
      comp opts: ComponentOptions {componentOptions = [], componentRoot = "C:\\Users\\runneradmin\\AppData\\Local\\Temp\\hie-bios-test-bd77a1b14cf6e038", componentDependencies = []} (0.24s)
      (1,2)                                                          (0.05s)
      (2,2)
```


Seems like the `shell` does not work for the string: 
```
ECHO "-Wall"
ECHO "A"
ECHO "B"
```

**Comment by pepeiborra** [comment]
Thanks for investigating! 
I have no Windows box at hand, only tried to extend the test with dependencies following the same pattern. 

Can the test be fixed at all? If not, what's the solution, disable it on windows?

**Comment by jneira** [comment]
> Can the test be fixed at all? If not, what's the solution, disable it on windows?

Well, at very least we could run two different test, one per os. Please dont disable it 😸

**Comment by fendor** [comment]
To me, it seems like the shell script feature just does not work for windows. Unfortunately, I dont know whether this can be fixed.

To be clear, the following does:
```haskell
cmd = concat
  [ "ECHO \"-Wall\"\r\n"
  , "ECHO \"A\"\r\n"
  , "ECHO \"B\"\r\n"
  ]

> runProcessWithExitCode (shell cmd) ""
(ExitSuccess,"\"-Wall\"\n","")
```

Second and third line are swallowed. Maybe we have to use something different to `shell`?

File: tests/projects/simple-bios-shell/hie.yaml [hunk 1]
```diff
@@ -8,3 +8,7 @@ cradle:
       ECHO "-Wall" >> %HIE_BIOS_OUTPUT%
       ECHO "A" >> %HIE_BIOS_OUTPUT%
       ECHO "B" >> %HIE_BIOS_OUTPUT%
+    dependency-shell: |
+      echo -n # > nul && ECHO hie.yaml >> %HIE_BIOS_OUTPUT%
```
Comment (jneira) [inline tests/projects/simple-bios-shell/hie.yaml:12]:
```suggestion
      echo -n # > nul && ECHO hie.yaml>> %HIE_BIOS_OUTPUT%
```
watching the logs only left this little step, thanks for take care 😃

**Comment by pepeiborra** [comment]
Can this be merged now?

**Review by jneira — APPROVED** [review]

**Comment by fendor** [comment]
Thank you for your patience, merged it now

**Comment by pepeiborra** [comment]
No worries, thank you and @jneira  for your help!

**Pull request merged**